src/tree/gpu_hist/evaluate_splits.cu

```diff
@@ -8,316 +8,269 @@
-              TempStorageT* temp_storage) {
-  __shared__ cub::Uninitialized<GradientSumT> uninitialized_sum;
-  GradientSumT& shared_sum = uninitialized_sum.Alias();
-
-  GradientSumT local_sum = GradientSumT();
-  // For loop sums features into one block size
-  auto begin = feature_histogram.data();
-  auto end = begin + feature_histogram.size();
-  for (auto itr = begin; itr < end; itr += BLOCK_THREADS) {
-    bool thread_active = itr + threadIdx.x < end;
-    // Scan histogram
-    GradientSumT bin = thread_active ? *(itr + threadIdx.x) : GradientSumT();
-    local_sum += bin;
-  }
-  local_sum = ReduceT(temp_storage->sum_reduce).Reduce(local_sum, cub::Sum());
-  // Reduction result is stored in thread 0.
-  if (threadIdx.x == 0) {
-    shared_sum = local_sum;
+  // Handle empty (trivial) input
+  if (l_n_features == 0 && r_n_features == 0) {
+    dh::LaunchN(out_splits.size(), [=]__device__(std::size_t idx) {
+      out_splits[idx] = DeviceSplitCandidate{};
+    });
+    return;
   }
-  cub::CTA_SYNC();
-  return shared_sum;
-}
 
-template <typename GradientSumT, typename TempStorageT> struct OneHotBin {
-  GradientSumT __device__ operator()(
-      bool thread_active, uint32_t scan_begin,
-      SumCallbackOp<GradientSumT>*,
-      GradientSumT const &missing,
-      EvaluateSplitInputs<GradientSumT> const &inputs, TempStorageT *) {
-    GradientSumT bin = thread_active
-                           ? inputs.gradient_histogram[scan_begin + threadIdx.x]
-                           : GradientSumT();
-    auto rest = inputs.parent_sum - bin - missing;
-    return rest;
+  auto out_scan = EvaluateSplitsGenerateSplitCandidatesViaScan(evaluator, left, right);
+  auto d_out_scan = dh::ToSpan(out_scan);
+
+  auto reduce_key = dh::MakeTransformIterator<int>(
+      thrust::make_counting_iterator<std::size_t>(0),
+      [d_out_scan] __device__(std::size_t i) {
+        return d_out_scan[i].node_idx;
+      });
+  auto reduce_val = dh::MakeTransformIterator<DeviceSplitCandidate>(
+      thrust::make_counting_iterator<std::size_t>(0),
+      [d_out_scan] __device__(std::size_t i) {
+        const auto& e = d_out_scan[i];
+        GradientSumT left_sum, right_sum;
+        if (e.is_cat) {
+          left_sum = e.parent_sum - e.partial_sum;
+          right_sum = e.partial_sum;
+        } else {
+          if (e.direction == DefaultDirection::kRightDir) {
+            left_sum = e.partial_sum;
+            right_sum = e.parent_sum - e.partial_sum;
+          } else {
+            left_sum = e.parent_sum - e.partial_sum;
+            right_sum = e.partial_sum;
+          }
+        }
+        return DeviceSplitCandidate{e.loss_chg, e.direction, e.findex, e.fvalue, e.is_cat,
+                                    GradientPair{left_sum}, GradientPair{right_sum}};
+      });
+  /**
+   * Perform segmented reduce to find the best split candidate per node.
+   * Note that there will be THREE segments:
```

Comment:
Why are there three segments? It looks like you're only using two output values.

Reply:
This is because we first generate candidate splits with a forward pass over the gradient histogram, in the order of ascending `node_idx` (0 or 1), and then generate candidate splits with a backward pass over the histgoram, in the order of descending `node_idx`. So the list of candidate splits generated will consist of three segments
```
[segment for left child node] [segment for right child node] [segment for left child node]
```